### PR TITLE
Convert JSDebugger Slider to a React SpeedSlider

### PR DESCRIPTION
**Pull request by caleybrock**
Initial work required for animation preview on GameLab.
- same functionality as original slider
- todo: share with turtle
- updated to use SpeedSlider in JSDebugger (Gamelab and Applab, but the Gamelab slider currently has no functionality)
- touch support
- added to our storybook

screen shot 2016-08-15 at 11 13 51 am

Note: @islemaster and I paired on most of this, thanks for your help!

cc @bcjordan - you mentioned you wanted to see some more React PRs :)

### CODE REVIEW

File: apps/src/JsDebuggerUi.js [hunk 1]
```diff
@@ -31,9 +32,10 @@ var WATCH_VALUE_NOT_RUNNING = "undefined";
  * App Lab, Game Lab, etc.
  * @param {!function} runApp - callback for "launching" the app, which is used
  *        by the "Step In" button when the app isn't running.
+ * @param {!ReduxStore} reduxStore
```
Comment (islemaster) [inline apps/src/JsDebuggerUi.js]: Update: Anyplace we used this `ReduxStore` type should just be `Store` (it's what [`createStore` is documented to return](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/redux.js#L17)).

File: apps/src/JsDebuggerUi.js [hunk 2]
```diff
@@ -150,19 +158,9 @@ JsDebuggerUi.prototype.initializeAfterDomCreated = function (options) {
       this.rootDiv_,
       document.getElementById('codeTextbox'));
 
-  // Initialize debug speed slider
-  var slider = this.rootDiv_.querySelector('#speed-slider');
-  if (slider) {
-    var sliderXOffset = 10,
-        sliderYOffset = 22,
-        sliderWidth = 130;
```
Comment (islemaster) [inline apps/src/JsDebuggerUi.js:158]: Does the new slider end up being 130px wide?

File: apps/src/templates/SpeedSlider.jsx [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    return Math.min(Math.max(val, 0), 1);
+  },
+
+  svgPositionToValue(position) {
+    position = (position - knobXMin)/TRACK_LENGTH;
+    return this.clampValue(position);
+  },
+
+  onTrackMouseDown(event) {
+    let mousePosition = this.mouseToSvg_(event);
```
Comment (bcjordan) [inline apps/src/templates/SpeedSlider.jsx]: Could this & other `let`s in this module be `const`s?

File: apps/src/JsDebuggerUi.js [hunk 4]
```diff
@@ -221,20 +219,15 @@ JsDebuggerUi.prototype.initializeAfterDomCreated = function (options) {
  * @return {number|undefined}
  */
 JsDebuggerUi.prototype.getStepDelay = function () {
-  if (this.speedSlider_) {
-    return JsDebuggerUi.stepDelayFromStepSpeed(this.speedSlider_.getValue());
-  }
-  return undefined;
```
Comment (bcjordan) [inline apps/src/JsDebuggerUi.js:227]: Awesome this all lets us get rid of these checks :)

File: apps/src/templates/SpeedSlider.jsx [hunk 5]
```diff
@@ -0,0 +1,233 @@
+var React = require('react');
+var color = require('../color');
+var Radium = require('radium');
+var dom = require('../dom');
```
Comment (islemaster) [inline apps/src/templates/SpeedSlider.jsx]: Nit: I'd be nice to use ES6-style imports for these, `import React from 'react';` etc.  Might need to be careful importing dom.js that way though, it's using nonstandard export syntax.

File: apps/src/templates/SpeedSlider.jsx [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    strokeWidth: 4,
+    strokeLinecap: "round",
+  },
+  sliderKnob: {
+    fill: "#ddd",
+    stroke: "#bbc",
+    strokeWidth: 1,
+    strokeLinejoin: "round",
+    ':hover': {
+      fill: "#eee"
```
Comment (islemaster) [inline apps/src/templates/SpeedSlider.jsx:18]: Can any of these colors be pulled from color.js?

**Comment by bcjordan** [comment]
(Very cool, thanks for the mention! :tada:)

File: apps/src/templates/SpeedSlider.jsx [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+    }
+  }
+};
+
+var sliderImages = {
```
Comment (islemaster) [inline apps/src/templates/SpeedSlider.jsx]: s/var/const

File: apps/src/templates/SpeedSlider.jsx [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+    this.isIOS_ = dom.isIOS();
+    this.isWindowsTouch_ = dom.isWindowsTouch();
+
+    var thisSlider = this;
+    dom.addMouseDownTouchEvent(this.knob_, function (e) {
+      return thisSlider.onKnobMouseDown(e);
+    });
+    dom.addMouseDownTouchEvent(this.track_, function (e) {
+      return thisSlider.onTrackMouseDown(e);
+    });
```
Comment (islemaster) [inline apps/src/templates/SpeedSlider.jsx]: React component methods are implicitly bound, and the inner callbacks here have the same contract as the functions you're creating - so you could remove the temporary `thisSlider` and just write:

``` javascript
dom.addMouseDownTouchEvent(this.knob_, this.onKnobMouseDown);
dom.addMouseDownTouchEvent(this.track_, this.onTrackMouseDown);
```

File: apps/src/templates/SpeedSlider.jsx [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+  },
+
+  stopDragging(event) {
+    this.setState(this.getInitialState());
+    this.unbindOnMouseMove();
+    this.unbindStopDragging();
+  },
+
+  render() {
+    const props = this.props;
```
Comment (islemaster) [inline apps/src/templates/SpeedSlider.jsx:134]: It'd be nice to clean this up.  Instead, let's deconstruct the props we need so we can use them as locals:

``` javascript
const {value, style, hasFocus} = this.props;
let clampedValue = this.clampValue(value);
// etc...
```

File: apps/src/templates/SpeedSlider.jsx [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+    let clampedValue = this.clampValue(props.value);
+    const knobWidth = 16;
```
Comment (islemaster) [inline apps/src/templates/SpeedSlider.jsx]: Could pull this constant up with the other constants?

File: apps/src/templates/SpeedSlider.jsx [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+    let knobXPosition = knobXMin + TRACK_LENGTH*clampedValue - 1/2*knobWidth;
```
Comment (islemaster) [inline apps/src/templates/SpeedSlider.jsx]: Can we simplify the end of this expression to `knobWidth/2`?

File: apps/src/templates/SpeedSlider.jsx [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+    return (
+      <div id="slider-cell" style={props.style}>
+        <svg
+          id="speed-slider"
```
Comment (islemaster) [inline apps/src/templates/SpeedSlider.jsx]: This might not need an ID anymore.

File: apps/src/templates/SpeedSlider.jsx [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+          version="1.1"
```
Comment (islemaster) [inline apps/src/templates/SpeedSlider.jsx]: I _think_ this version attribute isn't needed for SVG within React?

File: apps/src/templates/SpeedSlider.jsx [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+          <line
+            style={style.sliderTrack}
+            x1="10"
+            y1="25"
+            x2="140"
+            y2="25"
+            ref={el => this.track_ = el}
+          />
+          <path
+            id="knob"
```
Comment (islemaster) [inline apps/src/templates/SpeedSlider.jsx]: This might not need an ID either?

**Comment by islemaster** [comment]
_THIS ROCKS._

I've only got cleanup nits, really.  I think this looks like a sound change.  If you'd like we can walk through adding a unit test for the slider component using Enzyme too.

File: apps/src/JsDebuggerUi.js (hunk 2 above)
Comment (caleybrock) [inline apps/src/JsDebuggerUi.js:158]: Yep!

File: apps/src/templates/SpeedSlider.jsx (hunk 5 above)
Comment (caleybrock) [inline apps/src/templates/SpeedSlider.jsx]: I'm not getting errors when I `import dom from 'dom.js'` so I'll go ahead and move all of them over.

File: apps/src/templates/SpeedSlider.jsx (hunk 6 above)
Comment (caleybrock) [inline apps/src/templates/SpeedSlider.jsx:18]: unfortunately, I can't get any of the colours we have to look good here. Leaving as a one-off for now.

File: apps/src/templates/SpeedSlider.jsx (hunk 9 above)
Comment (caleybrock) [inline apps/src/templates/SpeedSlider.jsx:134]: I don't get any errors, but for some reason I'm having issues using `style` and `hasFocus` within the tags when doing it this way, so I might leave it as is for now.

**Comment by caleybrock** [comment]
@islemaster I'd love help with a test, but let's add a test when I add this to animation tab and give you time to be at a good stopping point with crypto widget.

All other comments are addressed - good to merge?

**Comment by islemaster** [comment]
LGTM!

**Pull request merged**